src/main/java/com/puppycrawl/tools/checkstyle/PackageObjectFactory.java

```diff
@@ -164,35 +179,36 @@ private Object createObjectFromMap(String name) throws CheckstyleException {
     }
 
     /**
-     * Create a new instance of a named class.
-     * @param className the name of the class to instantiate.
-     * @param secondAttempt the set of names to attempt instantiation
-     *                      if usage of the className was not successful.
-     * @return the {@code Object} created by loader or null.
-     * @throws CheckstyleException if the class fails to instantiate.
+     * Generate the map of third party Checkstyle module names to their fully qualified names.
+     * @param loader the class loader used to load Checkstyle package names
+     * @return the map of third party Checkstyle module names to their fully qualified names
      */
-    private Object createObjectWithIgnoringProblems(String className, Set<String> secondAttempt)
-            throws CheckstyleException {
-        Object instance = createObject(className);
-        if (instance == null) {
-            final Iterator<String> ite = secondAttempt.iterator();
-            while (instance == null && ite.hasNext()) {
-                instance = createObject(ite.next());
-            }
+    private Map<String, String> generateThirdPartyNameToFullModuleName(ClassLoader loader) {
+        Map<String, String> returnValue;
+        try {
+            final ClassPath classPath = ClassPath.from(loader);
+            returnValue = packages.stream().map(pkg -> pkg.substring(0, pkg.length() - 1))
+                    .flatMap(pkg -> classPath.getTopLevelClasses(pkg).stream())
+                    .map(ClassPath.ClassInfo::load)
+                    .filter(PackageObjectFactory::shouldAddToThirdPartyMap)
+                    .collect(Collectors.toMap(Class::getSimpleName, Class::getCanonicalName));
         }
-        return instance;
+        catch (IOException ignore) {
+            returnValue = new HashMap<>();
+        }
+        return returnValue;
     }
 
     /**
-     * Generate the set of all possible names for a class name.
-     * @param name the name of the class get possible names for.
-     * @return all possible name for a class.
+     * Whether the class should be added to the third party map.
+     * @param clazz the class to check
+     * @return true if the class should be added to the third party map
      */
-    private Set<String> getAllPossibleNames(String name) {
-        final Set<String> names = new HashSet<>();
-        names.addAll(packages.stream().map(packageName -> packageName + name)
-            .collect(Collectors.toList()));
-        return names;
+    private static boolean shouldAddToThirdPartyMap(Class<?> clazz) {
+        return !Modifier.isAbstract(clazz.getModifiers())
+                && !clazz.getSimpleName().startsWith("Input")
+                && (AbstractCheck.class.isAssignableFrom(clazz)
+                    || Filter.class.isAssignableFrom(clazz));
```

Comment:
Third party check can be anything of ours that can in configuration file.
This includes FileSets, BeforeExecutionFileFilters, RootModules, etc. which you don't list.

This should almost mimic `CheckUtil.isCheckstyleModule` which is why I mentioned why I think we should move it from test to production.